runtime/js-compute-runtime/builtins/shared/console.cpp

```diff
@@ -431,22 +434,390 @@ static bool console_out(JSContext *cx, unsigned argc, JS::Value *vp) {
+  }
+  if (fullLogLine.length()) {
+    fullLogLine += " ";
+    fullLogLine += source;
+  } else {
+    fullLogLine += source;
+  }
+
+  builtin_impl_console_log(Console::LogType::Log, fullLogLine.c_str());
+
+  args.rval().setUndefined();
+  return true;
+}
+
+static bool trace(JSContext *cx, unsigned argc, JS::Value *vp) {
+  JS::CallArgs args = CallArgsFromVp(argc, vp);
+  // 1. Let trace be some implementation-specific, potentially-interactive representation of the
+  // callstack from where this function was called.
+
+  JS::RootedObject stack(cx);
+  if (!CaptureCurrentStack(cx, &stack, JS::StackCapture(JS::MaxFrames(1u << 7)))) {
+    return false;
+  }
+  auto principals = JS::GetRealmPrincipals(JS::GetCurrentRealmOrNull(cx));
+  JS::RootedString str(cx);
+  if (!BuildStackString(cx, principals, stack, &str)) {
+    return false;
+  }
+  size_t length;
+  auto stack_string = encode(cx, str, &length);
+  if (!stack_string) {
+    return false;
+  }
+
+  // 2. Optionally, let formattedData be the result of Formatter(data), and incorporate
+  // formattedData as a label for trace.
+  std::string fullLogLine = "";
+  JS::RootedObjectVector visitedObjects(cx);
+  for (int i = 0; i < args.length(); i++) {
+    JS::HandleValue arg = args.get(i);
+    std::string source = "";
+    auto result = ToSource(cx, source, arg, &visitedObjects);
+    if (result.isErr()) {
+      return false;
+    }
+    // strip quotes for direct string logs
+    if (source[0] == '"' && source[source.length() - 1] == '"') {
+      source = source.substr(1, source.length() - 2);
+    }
+    if (fullLogLine.length()) {
+      fullLogLine += " ";
+      fullLogLine += source;
+    } else {
+      fullLogLine += source;
+    }
+  }
+  fullLogLine += "\n";
+  fullLogLine += stack_string.get();
+
+  // 3. Perform Printer("trace", « trace »).
+  builtin_impl_console_log(Console::LogType::Log, fullLogLine.c_str());
+  args.rval().setUndefined();
+  return true;
+}
+
 const JSFunctionSpec Console::methods[] = {
-    JS_FN("log", (console_out<Console::LogType::Log>), 1, JSPROP_ENUMERATE),
-    JS_FN("debug", (console_out<Console::LogType::Debug>), 1, JSPROP_ENUMERATE),
-    JS_FN("info", (console_out<Console::LogType::Info>), 1, JSPROP_ENUMERATE),
-    JS_FN("warn", (console_out<Console::LogType::Warn>), 1, JSPROP_ENUMERATE),
-    JS_FN("error", (console_out<Console::LogType::Error>), 1, JSPROP_ENUMERATE),
+    JS_FN("assert", assert, 0, JSPROP_ENUMERATE),
+    JS_FN("clear", no_op, 0, JSPROP_ENUMERATE),
+    JS_FN("count", count, 0, JSPROP_ENUMERATE),
+    JS_FN("countReset", countReset, 0, JSPROP_ENUMERATE),
+    JS_FN("debug", (console_out<Console::LogType::Debug>), 0, JSPROP_ENUMERATE),
+    JS_FN("dir", dir, 0, JSPROP_ENUMERATE),
+    JS_FN("dirxml", (console_out<Console::LogType::Log>), 0, JSPROP_ENUMERATE),
+    JS_FN("error", (console_out<Console::LogType::Error>), 0, JSPROP_ENUMERATE),
+    JS_FN("group", (console_out<Console::LogType::Log>), 0, JSPROP_ENUMERATE),
```

Comment:
Node.js and Deno have an indentation counter for grouping, although doesn't seem to be that valuable a feature to skip for now to me.